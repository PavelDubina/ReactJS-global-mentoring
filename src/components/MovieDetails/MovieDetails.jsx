import React, { useEffect } from 'react'
import { useParams, Link } from 'react-router-dom'
import { useSelector, useDispatch } from 'react-redux'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { faSearch } from '@fortawesome/free-solid-svg-icons'
import className from 'classnames'
import { ImageWithFallback } from '../ImageWithFallback/ImageWithFallback'
import { getReleaseYear } from '../../utils/helpers'
import fetchMovie from '../../redux/actions/fetchMovie'
import styles from './MovieDetails.scss'

export const MovieDetails = () => {
  const dispatch = useDispatch()
  const { id } = useParams()
  const { movie } = useSelector((state) => state.movieData)
  const runtimeStyle = className(styles.runtime, { [styles.noneVisible]: !movie.runtime })
  const ratingStyle = className(styles.rating, { [styles.noneVisible]: !movie.vote_average })
  useEffect(() => dispatch(fetchMovie(id)), [id])
  return (
    <div className={styles.container}>
      <ImageWithFallback src={movie.poster_path} id={id} />
      {movie.id ? (
        <div className={styles.info}>
          <div className={styles.name_and_rating}>
            <h1 className={styles.name}>{movie.title}</h1>
            <div className={ratingStyle}>{movie.vote_average}</div>
          </div>
          <div className={styles.duration_and_release}>
            <p>{getReleaseYear(movie.release_date)}</p>
            <p className={runtimeStyle}>
              {movie.runtime}
              {' min'}
            </p>
          </div>
          <p className={styles.overview}>{movie.overview}</p>
        </div>
      ) : (
        <p className={styles.not_found}>No movie found</p>
      )}
      <Link to="/">
        <FontAwesomeIcon className={styles.search_icon} icon={faSearch} />
      </Link>
    </div>
  )
}
