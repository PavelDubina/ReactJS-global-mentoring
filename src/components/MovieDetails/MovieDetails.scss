.container {
  position: relative;
  padding: 48px 0 0 56px;
  display: flex;
  height: 424px;
  width: 100%;
}

img {
  width: 240px;
  height: 340px;
}

.info {
  margin: 0 0 0 64px;
  display: flex;
  flex-direction: column;
  color: #fff;
}

.name_and_rating {
  display: flex;
  align-items: center;
}

.name {
  font-size: 44px;
  font-weight: normal;
}

.rating {
  margin: 0 0 0 16px;
  width: 60px;
  height: 60px;
  display: flex;
  align-items: center;
  justify-content: center;
  border: 1px solid #fff;
  border-radius: 50%;
  font-size: 28px;
  color: green;
}

.duration_and_release {
  padding: 12px 0;
  display: flex;
  font-size: 28px;
  color: #f65261;
}

.runtime {
  margin: 0 0 0 32px;
}

.noneVisible {
  display: none;
}

.overview {
  padding: 0 64px 0 0;
  font-size: 20px;
  overflow-y: auto;
}

.search_icon {
  position: absolute;
  right: 0;
  top: 0;
  color: #f65261;
  font-size: 24px;
  transform: rotate(90deg);
  cursor: pointer;
}

.not_found {
  flex: 1;
  text-align: center;
  align-self: center;
  font-size: 64px;
  color: #fff;
}
