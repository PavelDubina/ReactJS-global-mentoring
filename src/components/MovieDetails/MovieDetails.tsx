import React, { useEffect } from 'react'
import { useRouter } from 'next/router'
import { useSelector, useDispatch } from 'react-redux'
import className from 'classnames'
import { ImageWithFallback } from '../ImageWithFallback/ImageWithFallback'
import { getReleaseYear } from '../../utils/helpers'
import fetchMovie from '../../redux/actions/fetchMovie'
import icon from '../../../public/icon-search.svg'
import styles from '../../../styles/MovieDetails.module.scss'
import { TAppState } from '../../redux/reducers/rootReducer'

export const MovieDetails = () => {
  const router = useRouter()
  const { id } = router.query
  const dispatch = useDispatch()
  const { movie } = useSelector((state: TAppState) => state.movieData)
  const runtimeStyle = className(styles.runtime, { [styles.noneVisible]: !movie.runtime })
  const ratingStyle = className(styles.rating, { [styles.noneVisible]: !movie.vote_average })
  useEffect(() => {
    if (router.isReady) {
      dispatch(fetchMovie(String(id)))
    }
  }, [id])

  return (
    <div className={styles.container}>
      {movie.poster_path && <ImageWithFallback src={movie.poster_path} id={id} />}
      {movie.id ? (
        <div className={styles.info}>
          <div className={styles.name_and_rating}>
            <h1 className={styles.name}>{movie.title}</h1>
            <div className={ratingStyle}>{movie.vote_average}</div>
          </div>
          <div className={styles.duration_and_release}>
            <p>{getReleaseYear(movie.release_date)}</p>
            <p className={runtimeStyle}>
              {movie.runtime}
              {' min'}
            </p>
          </div>
          <p className={styles.overview}>{movie.overview}</p>
        </div>
      ) : (
        <p className={styles.not_found}>No movie found</p>
      )}
      <img className={styles.search_icon} onClick={() => router.push('/')} src={icon} alt="icon" />
    </div>
  )
}
