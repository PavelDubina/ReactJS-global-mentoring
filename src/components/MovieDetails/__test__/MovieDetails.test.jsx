import React from 'react'
import configureStore from 'redux-mock-store'
import { render } from '@testing-library/react'
import { Provider } from 'react-redux'
import movie from '../../../MockedData/mockMovie.json'
import { MovieDetails } from '../MovieDetails'

const mockStore = configureStore([])

jest.mock('next/router', () => ({ useRouter: jest.fn().mockReturnValue({ query: { id: '3133369' } }) }))

describe('MovieDetails', () => {
  it('Should renders correctly if movie found', () => {
    const store = mockStore({ movieData: { movie } })
    store.dispatch = jest.fn()
    const getComponent = () =>
      render(
        <Provider store={store}>
          <MovieDetails />
        </Provider>,
      )
    expect(getComponent({}).asFragment()).toMatchSnapshot()
  })
  it('Should renders correctly if movie not found', () => {
    const store = mockStore({ movieData: { movie: {} } })
    store.dispatch = jest.fn()

    const getComponent = () =>
      render(
        <Provider store={store}>
          <MovieDetails />
        </Provider>,
      )
    expect(getComponent({}).asFragment()).toMatchSnapshot()
  })
})
