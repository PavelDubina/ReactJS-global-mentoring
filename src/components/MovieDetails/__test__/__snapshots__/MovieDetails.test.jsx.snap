// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieDetails Should renders correctly if movie found 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <img
      alt="poster"
      height="16"
      src="https://image.tmdb.org/t/p/w500/ylXCdC106IKiarftHkcacasaAcb.jpg"
      width="9"
    />
    <div
      class="info"
    >
      <div
        class="name_and_rating"
      >
        <h1
          class="name"
        >
          La La Land
        </h1>
        <div
          class="rating"
        >
          7.9
        </div>
      </div>
      <div
        class="duration_and_release"
      >
        <p>
          2016
        </p>
        <p
          class="runtime"
        >
          128 min
        </p>
      </div>
      <p
        class="overview"
      >
        Mia, an aspiring actress, serves lattes to movie stars in between auditions and Sebastian, a jazz musician, scrapes by playing cocktail party gigs in dingy bars, but as success mounts they are faced with decisions that begin to fray the fragile fabric of their love affair, and the dreams they worked so hard to maintain in each other threaten to rip them apart.
      </p>
    </div>
    <a
      href="/"
    >
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-search fa-w-16 search_icon"
        data-icon="search"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 512 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z"
          fill="currentColor"
        />
      </svg>
    </a>
  </div>
</DocumentFragment>
`;

exports[`MovieDetails Should renders correctly if movie not found 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <img
      alt="poster"
      height="16"
      src="test-file-stub"
      width="9"
    />
    <p
      class="not_found"
    >
      No movie found
    </p>
    <a
      href="/"
    >
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-search fa-w-16 search_icon"
        data-icon="search"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 512 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z"
          fill="currentColor"
        />
      </svg>
    </a>
  </div>
</DocumentFragment>
`;
